etcdserver/snapshot_hub.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdserver
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/coreos/etcd/pkg/fileutil"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+type snapshot struct {
+	// storing the raft snapshot. the data in raftsnap is v2 store.
+	raftsnap raftpb.Snapshot
+	// rc is a reader that provides the data of KV snapshot.
+	rc io.ReadCloser
+	// dataSize specifies the size of snapshot data.
+	dataSize int64
+}
+
+type snapshotHub struct {
+	// dir to save snapshot data
+	dir string
+
+	// send empty to reqsnapc to notify the channel receiver to send back latest
+	// snapshot to snapc
+	reqsnapc chan struct{}
+	// a chan to receive the requested snapshot
+	// hub will receive from the chan immediately after it sends empty to reqsnapc
+	snapc chan snapshot
+	// host received snapshots
```

Comment:
To make things simpler from the very beginning, we probably should only support one out-going snapshot at a time.